packages/block-editor/src/components/line-height-control/with-line-height.js

```diff
@@ -0,0 +1,45 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getLineHeightControlStyles,
+	getLineHeightControlClassName,
+} from './utils';
+
+/**
+ * Higher-order component that provides a Block's Edit/Save component with
+ * interpolated className and style based on the lineHeight attribute value.
+ */
+export default function withLineHeight() {
```

Comment:
This is also a "new thing". using HoCs for "save" functions is a new pattern that raises some concern:

 - The implicitness/magic might not be a desired behavior
 - The difference with all the remaining attributes.
 - Lack of flexibility as it assumes that the className and styles across attributes are applied to the same element (not sure if it's always the case)

Reply:
> Lack of flexibility as it assumes that the className and styles across attributes are applied to the same element

I can see this one being a potential concern 🤔 
Do you have any suggestions for the other 2 points?

Reply:
I'd also strive to avoid magic in crucial places such as `save`. 

> Do you have any suggestions for the other 2 points? (With the goal of reducing verbosity in the `save` function)

It's clear that we need #20308 to yield some concept that we can try out regarding implicit attributes. I realise that React hooks currently cannot be used inside `save`, which further constrains us (and that's partly a good thing), otherwise I might say that we could solve this with a hook that would collect any implicit attributes that need to make their way into aspects of the serialised block.

An alternative I wonder about is how much we could leave to the server to deal with this instead, or how far we could take Andrew's idea of leveraging the unit of "the block" as a marker or carrier of style that is composable.

Reply:
> otherwise I might say that we could solve this with a hook...

@mcsf If only save worked with hooks 😊. As you pointed out, I went with the solution I did due to the save function not supporting hooks. The approach I took is similar in spirit with https://github.com/WordPress/gutenberg/pull/20308 and `useColor`, although perhaps not as daring. 

Ultimately, I think we need a systematic way to handle the setting/rendering of these style based attributes. It sounds like that is something we all agree upon 👍 . The question is when, where, and how.

This is something I've been chatting with @mtias about.

If folks are uncomfortable with the experimental abstraction at this strange, I'm happy to pull it back to follow the patterns we currently have.

Reply:
@ItsJonQ, thanks for indulging me.

> If folks are uncomfortable with the experimental abstraction at this strange, I'm happy to pull it back to follow the patterns we currently have.

I don't want this to block your exploration, so I'll leave this up to you, as long as the solution is something well contained that only applies to Paragraph (and hopefully doesn't mislead others in thinking they should follow a new design pattern).

My personal preference might be to, as you say, pull back and make the changes to `save` here as explicit and concrete as possible — leaving any abstractions to a stage at which we not only have better tooling but also broader understanding of what is needed. That said, you can still choose to encapsulate some of the functionality; if so, keep [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) in mind.

Reply:
Thanks for the feedback @mcsf !

I'll pull back the abstraction for now. Just a heads up, we're going to be [adding LineHeight controls for **Heading** as well](https://github.com/WordPress/gutenberg/issues/20339).

Perhaps after we remove the abstraction for Paragraph... We can revisit it (or something like it) when we add it to Heading.

Reply:
That makes sense to me.